inline-r/src/Language/R/Instance.hs

```diff
@@ -127,6 +131,19 @@ runRegion r =
              x <- runReaderT (unR r) d
              x `deepseq` return x)
 
+-- | Open a nested memory region. All objects allocated during the dynamic extent
+-- of a call to this function are liable to be freed by R's garbage collector
+-- once the call returns. Since the dynamic extent of a nested region is smaller
+-- than that of its parent region, using a nested regions allows objects to be
+-- freed sooner.
+withRegion :: forall a s. NFData a => (forall s'. (s' <= s) => R s' a) -> R s a
```

Comment:
The part `(forall s'. (s' <= s) => R s' a)` looks pretty scary for me, I'm happy that compiler is not confused and could infer types. This is not the solution that was proposed in Oleg's paper and that I was implementing, but it's nice and simple, and according to the tests it works.